### PR TITLE
Wifi camera model environment deterministic

### PR DESCRIPTION
**Pull request by HawkinH**
Environment-specific tests implemented. Matched indent spacing to original repo, though non-standard compliant. 

    $ SCARLET_TEST_HOST=192.168.0.100 SCARLET_TEST_PORT=80 SCARLET_TEST_WIFI=1 SCARLET_TEST_MODEL=bubl1 $(npm bin)/mocha

### CODE REVIEW

File: server_tests/test/clientTests.js [hunk 1]
```diff
@@ -13,11 +13,20 @@ var Q = require('../node_modules/q');
...
     var Comparison = new Compare();
     var Utility = new Util(testClient);
     var defaultOptionsFile = './defaults/mock.json';
     var timeoutValue = 30000;
+    var testViaWifi = process.env.SCARLET_TEST_WIFI == 1;
+    var testModel = process.env.SCARLET_TEST_MODEL;
+    var camModels = {
+      Bublcam1_0 : "bubl1",
+      Bublcam1_2 : "bubl2",
+      OtherCam : "osc"
```
Comment (AerialX) [inline server_tests/test/clientTests.js]: I would call this "OSC" or "Generic"

File: server_tests/test/clientTests.js [hunk 2]
```diff
@@ -308,7 +317,7 @@ describe("RUST API TEST SUITE", function() {
     });
 
     // CLOSE SESSION
-    describe("Testing /osc/commands/execute camera.closeSession endpoint", function() {
+    describe("Testing /osc/commands/execute camera.closeSession endpoint. Disabled for testing over Wifi", function() {
```
Comment (AerialX) [inline server_tests/test/clientTests.js]: "Disabled for" -> "Disabled when"

Comment (AerialX) [reply server_tests/test/clientTests.js]: I'm not fully convinced that we actually need to include this at all, though.

Comment (HawkinH) [reply server_tests/test/clientTests.js]: should i get rid of all of them?

Comment (AerialX) [reply server_tests/test/clientTests.js]: Er, this test in particular is not affected so this should definitely not be here. Apply the above comments to the setOptions wifiPassword test below.

Comment (HawkinH) [reply server_tests/test/clientTests.js]: also, the two tests are failing currently: bubl_Update and bubl_Shutdown

1) RUST API TEST SUITE Testing /osc/_bublUpdate endpoint. Bublcam exclusive test Expect success. /osc/_bublUpdate endpoint successfully returned status code 200:
     AssertionError: expected 500 to equal 200
      at verifyHttpStatus200 (lib/compare.js:659:16)
      at bublUpdateOutput (lib/compare.js:598:9)
      at test/clientTests.js:1900:28
      at lib/compare.js:687:24
      at _fulfilled (node_modules/q/q.js:834:54)
      at self.promiseDispatch.done (node_modules/q/q.js:863:30)
      at Promise.promise.promiseDispatch (node_modules/q/q.js:796:13)
      at node_modules/q/q.js:604:44
      at runSingle (node_modules/q/q.js:137:13)
      at flush (node_modules/q/q.js:125:13)
      at _combinedTickCallback (internal/process/next_tick.js:67:7)
      at process._tickCallback (internal/process/next_tick.js:98:9)

  2) RUST API TEST SUITE Testing /osc/commands/execute camera._bublShutdown. Bublcam exclusive test "before each" hook for "Expect success. camera._bublShutdown successfully returned when specific shutdownDelay is provided and returned at appropriate time. Disabled over wifi":
     TypeError: Cannot read property 'sessionId' of undefined
      at test/clientTests.js:1918:45
      at lib/compare.js:687:24
      at _fulfilled (node_modules/q/q.js:834:54)
      at self.promiseDispatch.done (node_modules/q/q.js:863:30)
      at Promise.promise.promiseDispatch (node_modules/q/q.js:796:13)
      at node_modules/q/q.js:604:44
      at runSingle (node_modules/q/q.js:137:13)
      at flush (node_modules/q/q.js:125:13)
      at _combinedTickCallback (internal/process/next_tick.js:67:7)
      at process._tickCallback (internal/process/next_tick.js:98:9)

File: server_tests/test/clientTests.js [hunk 3]
```diff
@@ -1490,7 +1509,12 @@ describe("RUST API TEST SUITE", function() {
     });
 
     // BUBL CAPTURE VIDEO
-    describe("Testing /osc/commands/execute camera._bublCaptureVideo endpoint", function() {
+    describe("Testing /osc/commands/execute camera._bublCaptureVideo endpoint. Bubl exclusive test", function() {
```
Comment (AerialX) [inline server_tests/test/clientTests.js]: I think for the Bubl specific tests we shouldn't modify the test title at all. It's pretty self-explanatory that the "_bubl" tests will not be run for generic cameras.

File: server_tests/test/clientTests.js [hunk 4]
```diff
@@ -1285,6 +1299,11 @@ describe("RUST API TEST SUITE", function() {
 
     // BUBL TIMELAPSE
     describe('Testing /osc/commands/execute camera._bublTimelapse command', function() {
+        if (!isBublcam) {
+          this.skip();
```
Comment (AerialX) [inline server_tests/test/clientTests.js]: Does this actually work? I recall from previous testing we couldn't call `skip()` from inside a `describe` block.

File: server_tests/test/clientTests.js [hunk 5]
```diff
@@ -1918,7 +1965,12 @@ describe("RUST API TEST SUITE", function() {
             }));
         });
 
-        it('Expect success. camera._bublShutdown successfully returned', function(done) {
+        it('Expect success. camera._bublShutdown successfully returned. Disabled over wifi', function(done) {
+            if (testViaWifi) {
```
Comment (AerialX) [inline server_tests/test/clientTests.js]: I'm thinking we need an additional parameter for this. Maybe a `testModel == camModels.BublcamMock`. Whether we're on wifi or not doesn't change the fact that the camera shutting down during the test is a problematic thing.

File: server_tests/test/clientTests.js (hunk 2 above)
Comment (AerialX) [reply server_tests/test/clientTests.js]: I'd probably just remove the title modifiers altogether. Especially the _bubl ones, and shutdown. The wifi ones I'm on the fence but leaning toward just keeping them out.

Comment (AerialX) [reply server_tests/test/clientTests.js]: The update test should indeed fail, and is a test that should only be executed for mock cameras, alongside shutdown.

**Comment by AerialX** [comment]
Feedback consolidation:

1. Remove the title modifiers, they're mostly unnecessary.
2. Add a model for our mock camera, called something like `BublcamMock` and `"mock"` or `"bublMock"` as the string parameter. This should also be included in the `isBublcam` check.
3. _bublUpdate is missing a check for mock.
4. shutdown should check for mock rather than wifi
5. is `skip()` correct inside a describe block?

File: server_tests/test/clientTests.js (hunk 2 above)
Comment (HawkinH) [reply server_tests/test/clientTests.js]: oh cool, that's good to know. so those two are running properly it seems.

File: server_tests/test/clientTests.js (hunk 4 above)
Comment (HawkinH) [reply server_tests/test/clientTests.js]: it worked for changing the wifi password... so i think it does?

File: server_tests/test/clientTests.js (hunk 5 above)
Comment (HawkinH) [reply server_tests/test/clientTests.js]: got it, will investigate and modify

File: server_tests/test/clientTests.js (hunk 4 above)
Comment (AerialX) [reply server_tests/test/clientTests.js]: wifi password is an `it()` block, not a `describe()` block. Did you run the tests with `MODEL=osc`, and they all pass / the correct ones are skipped?

File: server_tests/test/clientTests.js [hunk 6]
```diff
@@ -13,11 +13,22 @@ var Q = require('../node_modules/q');
 var Util = require('../lib/util');
 
 describe("RUST API TEST SUITE", function() {
-    var testClient = new OscClient(process.env.SCARLET_TEST_HOST, process.env.SCARLET_TEST_PORT);
+    var testClient = new OscClient(process.env.SCARLET_TEST_HOST, process.env.SCARLET_TEST_PORT, process.env.SCARLET_TEST_WIFI, process.env.SCARLET_TEST_MODEL);
```
Comment (jeffheifetz) [inline server_tests/test/clientTests.js]: OSC Client is NOT using this, so please remove

File: server_tests/test/clientTests.js [hunk 7]
```diff
@@ -13,11 +13,22 @@ var Q = require('../node_modules/q');
...
     var Comparison = new Compare();
     var Utility = new Util(testClient);
     var defaultOptionsFile = './defaults/mock.json';
     var timeoutValue = 30000;
+    var testViaWifi = process.env.SCARLET_TEST_WIFI == 1;
+    var testModel = process.env.SCARLET_TEST_MODEL;
```
Comment (jeffheifetz) [inline server_tests/test/clientTests.js]: Should probably be `var testModel = process.env.SCARLET_TEST_MODEL || camModels.MOCKCAM`

Comment (jeffheifetz) [reply server_tests/test/clientTests.js]: At least if we really do want the default to be mock as you implied in person

File: server_tests/test/clientTests.js [hunk 8]
```diff
@@ -1057,6 +1068,11 @@ describe("RUST API TEST SUITE", function() {
         });
 
         it("Expect success. camera.setOptions successfully sets options when wifiPassword option is set to supported value", function(done) {
+            if (testViaWifi) {
+                this.skip();
```
Comment (jeffheifetz) [inline server_tests/test/clientTests.js]: nitpick Should be `return this.skip();`

File: server_tests/test/clientTests.js [hunk 9]
```diff
@@ -13,11 +13,22 @@ var Q = require('../node_modules/q');
...
     var Utility = new Util(testClient);
     var defaultOptionsFile = './defaults/mock.json';
     var timeoutValue = 30000;
+    var testViaWifi = process.env.SCARLET_TEST_WIFI == 1;
+    var testModel = process.env.SCARLET_TEST_MODEL;
+    var camModels = {
+      BUBLCAM1_0 : "bubl1",
+      BUBLCAM1_2 : "bubl2",
+      OTHERCAM : "OSC",
+      MOCKCAM : "bublmock"
```
Comment (AerialX) [inline server_tests/test/clientTests.js]: 1. lowercase "osc" string
2. I'd really just call this MOCK rather than MOCKCAM. OTHERCAM would probably be better as OSC or GENERIC.

File: server_tests/test/clientTests.js (hunk 7 above)
Comment (AerialX) [reply server_tests/test/clientTests.js]: mm, probably should default to mock

File: server_tests/test/clientTests.js [hunk 10]
```diff
@@ -13,11 +13,22 @@ var Q = require('../node_modules/q');
 var Util = require('../lib/util');
 
 describe("RUST API TEST SUITE", function() {
-    var testClient = new OscClient(process.env.SCARLET_TEST_HOST, process.env.SCARLET_TEST_PORT);
+    var testClient = new OscClient(process.env.SCARLET_TEST_HOST, process.env.SCARLET_TEST_PORT, process.env.SCARLET_TEST_WIFI, process.env.SCARLET_TEST_MODEL);
     var Comparison = new Compare();
     var Utility = new Util(testClient);
     var defaultOptionsFile = './defaults/mock.json';
     var timeoutValue = 30000;
+    var testViaWifi = process.env.SCARLET_TEST_WIFI == 1;
```
Comment (AerialX) [inline server_tests/test/clientTests.js]: If we're talking defaults, this logic should probably be `!mock && wifi == 1`

File: server_tests/test/clientTests.js (hunk 9 above)
Comment (AerialX) [reply server_tests/test/clientTests.js]: (that or BUBLMOCK to indicate that it supports the bubl proprietary commands)

File: server_tests/test/clientTests.js [hunk 11]
```diff
@@ -1303,7 +1302,7 @@ describe("RUST API TEST SUITE", function() {
     describe('Testing /osc/commands/execute camera._bublTimelapse command', function() {
         before(function() {
             if (!isBublcam) {
-              this.skip();
+              return this.skip();
```
Comment (AerialX) [inline server_tests/test/clientTests.js]: this return doesn't do anything, but this check and before should be merged with the one under it instead (where the return **will** be important).

File: .gitignore [hunk 12]
```diff
@@ -0,0 +1 @@
+node_modules/
```
Comment (jeffheifetz) [inline .gitignore]: Turns out we don't need this, you just had node_modules in a weird place

File: server_tests/test/clientTests.js [hunk 13]
```diff
@@ -1069,7 +1069,7 @@ describe("RUST API TEST SUITE", function() {
 
         it("Expect success. camera.setOptions successfully sets options when wifiPassword option is set to supported value", function(done) {
             if (testViaWifi) {
-                return this.skip();
+                this.skip();
```
Comment (AerialX) [inline server_tests/test/clientTests.js]: this change is wrong, the return should be here

File: server_tests/test/clientTests.js [hunk 14]
```diff
@@ -1514,7 +1517,7 @@ describe("RUST API TEST SUITE", function() {
     describe("Testing /osc/commands/execute camera._bublCaptureVideo endpoint", function() {
         before(function() {
             if (!isBublcam) {
-                return this.skip();
+                this.skip();
```
Comment (AerialX) [inline server_tests/test/clientTests.js]: all of these other `before()`s are the same thing, should be merged as you did with timelapse.

File: OscClient.js [hunk 15]
```diff
@@ -63,7 +63,7 @@ var Q;
         var stateUrl = serverAddress + '/osc/state';
 
         this.getState = function() {
-            return makeHttpRequest('POST', stateUrl, applicationJsonType); 
+            return makeHttpRequest('POST', stateUrl, applicationJsonType);
```
Comment (jeffheifetz) [inline OscClient.js]: please revert this change to keep it minimal

File: server_tests/test/clientTests.js [hunk 16]
```diff
@@ -1851,6 +1890,12 @@ describe("RUST API TEST SUITE", function() {
 
     // BUBL UPDATE
     describe('Testing /osc/_bublUpdate endpoint', function() {
+        before(function() {
+            if (!isMock) {
```
Comment (arcnmx) [inline server_tests/test/clientTests.js]: This also needs to check isBublcam.
